prost-derive/src/lib.rs

```diff
@@ -412,23 +414,28 @@ fn try_oneof(input: TokenStream) -> Result<TokenStream, Error> {
         }
     }
 
-    let mut tags = fields
+    for (variant_ident, field) in &fields {
+        // Not clear if this condition is reachable since multiple "tag" attributes are already
+        // rejected, but good to be safe
```

Comment:
I don't like to leave this comment here. We should either:
- make sure this can't happen by validating the input, then this check can be replaced by an assertion, or
- verify that multiple tags are possible, but not allowed for oneof, then this code is good, and the comment can be removed.

It seems like the origin of the Vec is here. Can you trigger that code in a test? https://github.com/tokio-rs/prost/blob/993a6d132187a95c624bab527440e280e2f58af8/prost-derive/src/field/oneof.rs#L36

Reply:
nope, because that's the origin for the `tags = "1, 2, 3"` attribute that appears on `oneof` members *where they appear in a message*. This case we are testing is for the *variant* fields *within* a oneof enum, which is populated here:
https://github.com/tokio-rs/prost/blob/02bc13232e5a3d9aa041e8657456ea8590992863/prost-derive/src/field/scalar.rs#L34

...and elsewhere, the same as normal scalar/map/group/message fields are parsed. dispatch occurs from up above in [field/mod.rs](https://github.com/tokio-rs/prost/blob/04be6049ea676dde5dfac8a21da2559f4f380ed2/prost-derive/src/field/mod.rs#L66-L73), and the difference between `new()` and `new_oneof()` in these cases is at most some extra validation. The only kind of field we can parse that is allowed to have multiple tags is a oneof, which is parsed via `field::oneof::Field::new()`; there is no corresponding `::new_oneof()` function, and we do not have a branch for it in the upper dispatch. Naturally this is because oneof fields cannot themselves contain other oneofs.

Since all this is kind of obnoxious and long-winded to prove, and i don't hate the assertion, i'm happy with including it in whatever way we feel fits best. I'd reduced it from the full error message we see here to a simpler `panic!` originally; we can do that again. perhaps it's semantically clearer to express that as an `assert!`: that way it can stand in the code to provide context on the operating conditions of the surrounding code, even if we're already confident it's unreachable.

```rust
assert!(fields.iter().all(|field| field.tags().len() == 1));
```
...perhaps with a brief appropriate comment about how oneofs can't directly nest.

Reply:
well i spelled it wrong in that comment but i got it right in the end.

Reply:
Thanks